crates/matrix-sdk-crypto/src/types/one_time_keys.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright 2021 The Ruma contributors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//! Module containing customized types modeling Matrix one-time keys.
+//!
+//! These types were mostly taken from the Ruma project. The types differ in two
+//! important ways to the Ruma types of the same name:
+//!
+//! 1. They are using vodozemac types so we directly deserialize into a
+//!    vodozemac curve25519 key.
+//! 2. They support lossless serialization cycles.
+
+use std::collections::BTreeMap;
+
+use ruma::{serde::Raw, DeviceKeyId, UserId};
+use serde::{Deserialize, Serialize, Serializer};
+use serde_json::{value::to_raw_value, Value};
+use vodozemac::{Curve25519PublicKey, Ed25519Signature};
+
+/// Signatures for a `SignedKey` object.
+pub type SignedKeySignatures = BTreeMap<Box<UserId>, BTreeMap<Box<DeviceKeyId>, Ed25519Signature>>;
+
+/// A key for the SignedCurve25519 algorithm
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SignedKey {
+    // /// The Curve25519 key that can be used to establish Olm sessions.
+    #[serde(deserialize_with = "deserialize_curve_key", serialize_with = "serialize_curve_key")]
+    key: Curve25519PublicKey,
+
+    /// Signatures for the key object.
+    #[serde(deserialize_with = "deserialize_signatures", serialize_with = "serialize_signatures")]
+    signatures: SignedKeySignatures,
+
+    /// Is the key considered to be a fallback key.
+    #[serde(default, skip_serializing_if = "Option::is_none", deserialize_with = "double_option")]
+    fallback: Option<Option<bool>>,
```

Comment:
Maybe use [`js_option`](https://docs.rs/js_option/0.1.0/js_option/) here?

Reply:
Eh, I'd rather not add yet another dep for this, it's simple enough.